File: backend/src/modules/budgets/domain/repositories/budget.repository.ts
import { Budget, BudgetPeriod } from '../entities/budget.entity';

export abstract class BudgetRepository {
  /** Non-soft-deleted budgets owned by the user. */
  abstract findAllForUser(userId: string): Promise<Budget[]>;

  abstract findById(id: string): Promise<Budget | null>;

  /** Used for the uq_budgets_user_category_period pre-check (a friendlier
   * 409 CONFLICT message than surfacing the raw DB constraint error). */
  abstract findByUserCategoryPeriodAndStart(
    userId: string,
    categoryId: string,
    period: BudgetPeriod,
    startDate: string,
  ): Promise<Budget | null>;

  abstract create(data: {
    userId: string;
    categoryId: string;
    amountLimit: string;
    period: BudgetPeriod;
    startDate: string;
  }): Promise<Budget>;

  abstract update(
    id: string,
    changes: { amountLimit?: string; period?: BudgetPeriod },
  ): Promise<Budget>;

  abstract softDelete(id: string): Promise<void>;
}

File: mobile/lib/features/budgets/domain/repositories/budget_repository.dart
import '../entities/budget.dart';

abstract class BudgetRepository {
  /// GET /budgets — every budget owned by the caller, with progress
  /// computed server-side. Not paginated (docs/08_API.md §11).
  Future<List<Budget>> fetchAll();

  /// POST /budgets — 409s if a budget already exists for this
  /// categoryId/period/startDate (uq_budgets_user_category_period, T6.1).
  Future<Budget> create({
    required String categoryId,
    required String amountLimit,
    required BudgetPeriod period,
    required String startDate,
  });
}
